### PR TITLE
feat(status): add the status module 

### LINKED ISSUE #193: Option to show job exit status in prompt
#### Is your feature request related to a problem? Please describe.
When a job fails, starship displays the prompt character in red, but doesn't show me what the exit status was. This is a fairly subtle display, and the exit code can be useful information.

#### Describe the solution you'd like
I'd like a `directory` option that adds the status code. My current prompt does this by putting e.g. `[3]` (in red) right before the `>`. I'd like something similar for starship.

My Bash prompt actually goes a step further and shows me the entire pipeline output if any command had a non-zero exit, even if the overall status was success (in that case it's colored grey instead of red). So for the Bash prompt `false | true` will display as `[1 0]` (in grey), `true | false` will display as `[0 1]` (in red) and longer jobs will add more numbers. And if the overall job status doesn't match the final command status (e.g. because I used `!`) it tacks the overall status on the end in parens, so e.g. `! false` becomes `[1 (0)]` (in grey).

The only reason my Fish prompt doesn't do what my Bash prompt does is because Fish doesn't (yet) have a `$pipestatus` variable.

I know this is kind of complicated, but it works really well for me, especially when putting together complex command-lines, and I'd love to see the ability to do this with starship.

---

<details><summary>Bash <code>__prompt_command</code> function that implements this</summary>

```bash
__prompt_command () 
{ 
    local status=$? pipestatus=("${PIPESTATUS[@]}");
    local prompt="";
    prompt+='\[\e[1;32m\]\u\[\e[m\]@\[\e[1;36m\]\h\[\e[m\]:\[\e[1;31m\]';
    prompt+='$(__prompt_path)';
    prompt+='\[\e[0;33m\]';
    prompt+='$(__git_ps1 " (%s)")';
    prompt+='\[\e[m\]';
    local promptstatus=;
    if (( status > 255 )); then
        promptstatus=$status;
    else
        local x;
        for x in "${pipestatus[@]}";
        do
            if (( x )); then
                promptstatus=${pipestatus[*]};
                break;
            fi;
        done;
        if (( status != pipestatus[${#pipestatus[@]}-1] )); then
            promptstatus+="${promptstatus:+ }($status)";
        fi;
    fi;
    if [[ -n $promptstatus ]]; then
        local color=31;
        if (( status == 0 )); then
            color=2;
        fi;
        prompt+=" \\[\\e[${color}m\\][$promptstatus]\\[\\e[m\\]";
    fi;
    prompt+='> ';
    PS1=$prompt
}
```

</details>

### LINKED ISSUE #882: Support for status codes
## Feature Request

Requesting better support for status codes:
- The ability to see the specific (nonzero) exit code
- Different behaviors for "known special" codes (signals, 127, etc)

#### Describe the solution you'd like

Add a module that, at the very least, shows the code in (by default) red, when it's (by default) nonzero. 

#### Describe alternatives you've considered

None; this feature seems pretty straightforward.

### LINKED ISSUE #1641: show the exit code of the last command executed if it failed
## Feature Request

Screenshot from 2020-09-03 19-13-35

I do this with a custom module #1639

### PR DESCRIPTION
**Pull request by jankatins**
Replacement for the original PR in #1165 by @glehmann updated with all the changes from master and adjusted re best practises  (as I understand them)  for `format`, `symbol` and `style` usage.

#### Description

A "status" module which shows the exit code of the last command

#### Motivation and Context
Closes #193
Closes #882
Closes #1481
Closes #1641 

(and the original PR: Closes: #1140 )

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: docs/config/README.md [hunk 1]
```diff
@@ -2082,6 +2083,45 @@ The module will be shown if any of the following conditions are met:
...
 
+## Status
+
+The `status` module displays the exit code of the previous command.
+The module will be shown only if the exit code is not `0`.
+
+### Options
+
+| Variable   | Default                       | Description                                            |
+| ---------- | ----------------------- ------| ------------------------------------------------------ |
```
Comment (charego) [inline docs/config/README.md:2094]: The formatting for "Options" is messed up, probably due to the extra space.

format

status

**Review by charego — CHANGES_REQUESTED** [review]

**Comment by charego** [comment]
Tested on MacOS and it seems to work, thanks! I would really like to see the pipestatus added eventually.

Compare the default fish prompt:

pipestatus

With this feature's behavior:

status

File: docs/config/README.md (hunk 1 above)
Comment (jankatins) [reply docs/config/README.md:2094]: Fixed the out-of-space-space and missing space in the end, but the spacing in the symbol line seems to come from the chosen character and how big it is displayed in a specific font:

diff on the commandline:
image

vs 

jetbrains
image

vs 
gitkraken

image

**Comment by jankatins** [comment]
If I understand pipestatus right this needs another commandline argument (to pass in a string for all the exit statuses) and changes to the zsh/bash integrations (to pass in the variable from these two shells which have that feature)? https://unix.stackexchange.com/questions/14270/get-exit-status-of-process-thats-piped-to-another

File: docs/config/README.md [hunk 2]
```diff
@@ -2082,6 +2083,45 @@ The module will be shown if any of the following conditions are met:
 format = "via [🏎  $version](red bold)"
 ```
 
+## Status
+
+The `status` module displays the exit code of the previous command.
+The module will be shown only if the exit code is not `0`.
+
```
Comment (andytom) [inline docs/config/README.md:2090]: Let's add a tip here to make it explicitly obvious that this module is disabled by default
```suggestion

::: tip
                                                              
This module is disabled by default.
To enable it, set `disabled` to `false` in your configuration file.
:::

```

**Review by andytom — COMMENTED** [review]
Have added a suggestion to the docs. I'm wondering if the changes to the other modules, the `hg_branch` in particular might break anything for our users.

**Comment by jankatins** [comment]
> I'm wondering if the changes to the other modules, the hg_branch in particular might break anything for our users maybe those changes should be done in a separate PR?

In theory, the only case it could break for someone if that person uses the default config (=no changes to the hg module) and is in a hg repo and that person expects a hg output.

The only non-breaking change you could make is setting the default `disabled=true` to `disabled=false` to match the status quo.

It's currently a bug: e.g. after reading the docs, I would have expected that I won't see a hg output even if I have a hg repo clone and I am in that dir.

All the other modules are not effected, I only pasted in the same comment.

Your call: should I tke it out?

**Comment by jankatins** [comment]
For the pipeline status, I opened https://github.com/starship/starship/issues/1669

**Comment by jankatins** [comment]
Fixed the merge conflict

**Comment by andytom** [comment]
> > I'm wondering if the changes to the other modules, the hg_branch in particular might break anything for our users maybe those changes should be done in a separate PR?
> 
> In theory, the only case it could break for someone if that person uses the default config (=no changes to the hg module) and is in a hg repo and that person expects a hg output.
> 
> The only non-breaking change you could make is setting the default `disabled=true` to `disabled=false` to match the status quo.
> 
> It's currently a bug: e.g. after reading the docs, I would have expected that I won't see a hg output even if I have a hg repo clone and I am in that dir.
> 
> All the other modules are not effected, I only pasted in the same comment.
> 
> Your call: should I take it out?

My preference would be to remove the changes to the other modules from this PR and open a new PR for them. Since we auto-generate our release notes based on the squashed and merged PRs commit title, if we leave them in here the changes to the other modules won't show up in the changelog.

**Comment by jankatins** [comment]
> My preference would be to remove the changes to the other modules from this PR and open a new PR for them.

removed the unrelated fixes and will open a new PR for them

**Review by andytom — APPROVED** [review]

**Comment by andytom** [comment]
@jankatins Thank you for your contribution and thank you @glehmann for your original work on this.

**Pull request merged**